File: src/CSharp/144.BinaryTreePreorderTraversal.cs

using System;
using System.Text;
using System.Collections.Generic;

namespace PeefyLeetCode.PreorderTraversal
{

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}


public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        List<int> output = new List<int>();
        if (root == null) {
            return output;
        }

        stack.Push(root);
        while (!(stack.Count == 0)) {
            TreeNode node = stack.Pop();
            output.Add(node.val);
            if (node.right != null) {
                stack.Push(node.right);
            }
            if (node.left != null) {
                stack.Push(node.left);
            }
        }
        return output;
    }
}

}

File: src/Python/101-200/144.BinaryTreePreorderTraversal.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return []   
        stack, output = [root, ], []
        while stack:
            root = stack.pop()
            if root is not None:
                output.append(root.val)
                if root.right is not None:
                    stack.append(root.right)
                if root.left is not None:
                    stack.append(root.left)
        return output


if __name__ == "__main__":
    solution = Solution()
    node = TreeNode(2)
    node.left = TreeNode(1)
    node.right = TreeNode(3)
    node.right.left = TreeNode(4)
    print(solution.levelOrder(node))
   